# Ex1 - Filtering and Sorting Data

This time we are going to pull data directly from the internet.
Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### Step 1. Import the necessary libraries

In [1]:
import pandas as pd

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv). 

### Step 3. Assign it to a variable called chipo.

In [2]:
df_chipo=pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv", sep="\t")

In [50]:
df_chipo

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

### Step 4. How many products cost more than $10.00?

In [24]:
df_chipo.dtypes

order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
dtype: object

In [27]:
df_price=df_chipo["item_price"]

In [28]:
df_price

0        2.39
1        3.39
2        3.39
3        2.39
4       16.98
        ...  
4617    11.75
4618    11.75
4619    11.25
4620     8.75
4621     8.75
Name: item_price, Length: 4622, dtype: float64

In [32]:
df_chipo[df_price>10].count()

order_id              1130
quantity              1130
item_name             1130
choice_description    1123
item_price            1130
dtype: int64

In [33]:
df_chipo[df_price>10].nunique().count()

5

#### Porque pongo "." y no "chipo=pd."????

In [13]:
is_more= df_chipo["item_price"]> 10.00
print(is_more.head())

TypeError: '>' not supported between instances of 'str' and 'float'

In [14]:
df_chipo.dtypes["item_price"]

dtype('O')

In [15]:
df_chipo.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

#### Porque se pone \ antes del $???


In [20]:
df_chipo.item_price= df_chipo.item_price.replace("\$", "", regex=True)

In [21]:
df_chipo

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN      2.39   
1                                          [Clementine]      3.39   
2                                               [Apple]      3.39   
3                                                   NaN      2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...     16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...     11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...     11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...      8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...      8.75   

[4622 rows x 5 columns]

In [23]:
df_chipo.item_price=pd.to_numeric(df_chipo.item_price)

In [17]:
df_chipo.dtypes

order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
dtype: object

In [18]:
is_more=df_chipo["item_price"]>10

In [19]:
is_more.head()

0    False
1    False
2    False
3    False
4     True
Name: item_price, dtype: bool

In [27]:
is_more.value_counts()

False    3492
True     1130
Name: item_price, dtype: int64

### Step 5. What is the price of each item? 
###### print a data frame with only two columns item_name and item_price

In [35]:
df_chipo["item_name"]

0                Chips and Fresh Tomato Salsa
1                                        Izze
2                            Nantucket Nectar
3       Chips and Tomatillo-Green Chili Salsa
4                                Chicken Bowl
                        ...                  
4617                            Steak Burrito
4618                            Steak Burrito
4619                       Chicken Salad Bowl
4620                       Chicken Salad Bowl
4621                       Chicken Salad Bowl
Name: item_name, Length: 4622, dtype: object

In [36]:
print(df_chipo[["item_name","item_price"]])

                                  item_name  item_price
0              Chips and Fresh Tomato Salsa        2.39
1                                      Izze        3.39
2                          Nantucket Nectar        3.39
3     Chips and Tomatillo-Green Chili Salsa        2.39
4                              Chicken Bowl       16.98
...                                     ...         ...
4617                          Steak Burrito       11.75
4618                          Steak Burrito       11.75
4619                     Chicken Salad Bowl       11.25
4620                     Chicken Salad Bowl        8.75
4621                     Chicken Salad Bowl        8.75

[4622 rows x 2 columns]


In [40]:
df_unique= df_chipo.item_name.unique()

In [41]:
print(df_unique)

['Chips and Fresh Tomato Salsa' 'Izze' 'Nantucket Nectar'
 'Chips and Tomatillo-Green Chili Salsa' 'Chicken Bowl' 'Side of Chips'
 'Steak Burrito' 'Steak Soft Tacos' 'Chips and Guacamole'
 'Chicken Crispy Tacos' 'Chicken Soft Tacos' 'Chicken Burrito'
 'Canned Soda' 'Barbacoa Burrito' 'Carnitas Burrito' 'Carnitas Bowl'
 'Bottled Water' 'Chips and Tomatillo Green Chili Salsa' 'Barbacoa Bowl'
 'Chips' 'Chicken Salad Bowl' 'Steak Bowl' 'Barbacoa Soft Tacos'
 'Veggie Burrito' 'Veggie Bowl' 'Steak Crispy Tacos'
 'Chips and Tomatillo Red Chili Salsa' 'Barbacoa Crispy Tacos'
 'Veggie Salad Bowl' 'Chips and Roasted Chili-Corn Salsa'
 'Chips and Roasted Chili Corn Salsa' 'Carnitas Soft Tacos'
 'Chicken Salad' 'Canned Soft Drink' 'Steak Salad Bowl'
 '6 Pack Soft Drink' 'Chips and Tomatillo-Red Chili Salsa' 'Bowl'
 'Burrito' 'Crispy Tacos' 'Carnitas Crispy Tacos' 'Steak Salad'
 'Chips and Mild Fresh Tomato Salsa' 'Veggie Soft Tacos'
 'Carnitas Salad Bowl' 'Barbacoa Salad Bowl' 'Salad' 'Veggie Cris

In [62]:
df_ok= df_chipo[["item_name","item_price"]]

In [56]:
df_ok.nunique()

item_name     50
item_price    78
dtype: int64

#### es posible que haya mas item_name que item_price???

In [60]:
print(df_ok)

                                  item_name item_price
0              Chips and Fresh Tomato Salsa     $2.39 
1                                      Izze     $3.39 
2                          Nantucket Nectar     $3.39 
3     Chips and Tomatillo-Green Chili Salsa     $2.39 
4                              Chicken Bowl    $16.98 
...                                     ...        ...
4617                          Steak Burrito    $11.75 
4618                          Steak Burrito    $11.75 
4619                     Chicken Salad Bowl    $11.25 
4620                     Chicken Salad Bowl     $8.75 
4621                     Chicken Salad Bowl     $8.75 

[4622 rows x 2 columns]


In [34]:
df_name=df_chipo["item_name"].nunique()

In [35]:
df_name

50

### Step 6. Sort by the name of the item

In [72]:
df_chipo.sort_values(by="item_name")

order_id  quantity          item_name  \
3389      1360         2  6 Pack Soft Drink   
341        148         1  6 Pack Soft Drink   
1849       749         1  6 Pack Soft Drink   
1860       754         1  6 Pack Soft Drink   
2713      1076         1  6 Pack Soft Drink   
...        ...       ...                ...   
2384       948         1  Veggie Soft Tacos   
781        322         1  Veggie Soft Tacos   
2851      1132         1  Veggie Soft Tacos   
1699       688         1  Veggie Soft Tacos   
1395       567         1  Veggie Soft Tacos   

                                     choice_description item_price  
3389                                        [Diet Coke]    $12.98   
341                                         [Diet Coke]     $6.49   
1849                                             [Coke]     $6.49   
1860                                        [Diet Coke]     $6.49   
2713                                             [Coke]     $6.49   
...                                                 ...        ...  
2384  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75   
781   [Fresh Tomato Salsa, [Black Beans, Cheese, Sou...     $8.75   
2851  [Roasted Chili Corn Salsa (Medium), [Black Bea...     $8.49   
1699  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $11.25   
1395  [Fresh Tomato Salsa (Mild), [Pinto Beans, Rice...     $8.49   

[4622 rows x 5 columns]

#### Es correcto?

### Step 7. What was the quantity of the most expensive item ordered?

In [67]:
df_chipo.sort_values(by="item_price",ascending=False)

order_id  quantity            item_name  \
2624      1042         1     Steak Salad Bowl   
4419      1762         1     Steak Salad Bowl   
4036      1615         1     Steak Salad Bowl   
1825       738         1  Barbacoa Salad Bowl   
3115      1243         1  Carnitas Salad Bowl   
...        ...       ...                  ...   
3145      1254         1          Canned Soda   
414        180         1          Canned Soda   
3162      1262         1          Canned Soda   
821        338         1          Canned Soda   
1457       591         1          Canned Soda   

                                     choice_description item_price  
2624  [Fresh Tomato Salsa, [Black Beans, Sour Cream,...     $9.39   
4419  [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $9.39   
4036  [Fresh Tomato Salsa, [Fajita Vegetables, Chees...     $9.39   
1825  [Fresh Tomato Salsa, [Rice, Pinto Beans, Chees...     $9.39   
3115  [Tomatillo Green Chili Salsa, [Rice, Pinto Bea...     $9.39   
...                                                 ...        ...  
3145                                  [Diet Dr. Pepper]     $1.09   
414                                        [Dr. Pepper]     $1.09   
3162                                        [Coca Cola]     $1.09   
821                                         [Coca Cola]     $1.09   
1457                                           [Sprite]     $1.09   

[4622 rows x 5 columns]

#### algo no es correcto, porque me dice que el precio mas caro es $9.39????

In [69]:
df_caro=df_chipo.set_index("item_price")

In [70]:
df_caro

order_id  quantity                              item_name  \
item_price                                                              
$2.39              1         1           Chips and Fresh Tomato Salsa   
$3.39              1         1                                   Izze   
$3.39              1         1                       Nantucket Nectar   
$2.39              1         1  Chips and Tomatillo-Green Chili Salsa   
$16.98             2         2                           Chicken Bowl   
...              ...       ...                                    ...   
$11.75          1833         1                          Steak Burrito   
$11.75          1833         1                          Steak Burrito   
$11.25          1834         1                     Chicken Salad Bowl   
$8.75           1834         1                     Chicken Salad Bowl   
$8.75           1834         1                     Chicken Salad Bowl   

                                           choice_description  
item_price                                                     
$2.39                                                     NaN  
$3.39                                            [Clementine]  
$3.39                                                 [Apple]  
$2.39                                                     NaN  
$16.98      [Tomatillo-Red Chili Salsa (Hot), [Black Beans...  
...                                                       ...  
$11.75      [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...  
$11.75      [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...  
$11.25      [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...  
$8.75       [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...  
$8.75       [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...  

[4622 rows x 4 columns]

In [73]:
df_caro.sort_index(ascending=False)

order_id  quantity            item_name  \
item_price                                            
$9.39           1042         1     Steak Salad Bowl   
$9.39           1762         1     Steak Salad Bowl   
$9.39           1615         1     Steak Salad Bowl   
$9.39            738         1  Barbacoa Salad Bowl   
$9.39           1243         1  Carnitas Salad Bowl   
...              ...       ...                  ...   
$1.09           1254         1          Canned Soda   
$1.09            180         1          Canned Soda   
$1.09           1262         1          Canned Soda   
$1.09            338         1          Canned Soda   
$1.09            591         1          Canned Soda   

                                           choice_description  
item_price                                                     
$9.39       [Fresh Tomato Salsa, [Black Beans, Sour Cream,...  
$9.39       [Roasted Chili Corn Salsa, [Fajita Vegetables,...  
$9.39       [Fresh Tomato Salsa, [Fajita Vegetables, Chees...  
$9.39       [Fresh Tomato Salsa, [Rice, Pinto Beans, Chees...  
$9.39       [Tomatillo Green Chili Salsa, [Rice, Pinto Bea...  
...                                                       ...  
$1.09                                       [Diet Dr. Pepper]  
$1.09                                            [Dr. Pepper]  
$1.09                                             [Coca Cola]  
$1.09                                             [Coca Cola]  
$1.09                                                [Sprite]  

[4622 rows x 4 columns]

In [6]:
df_chipo["item_price"].max()

'$9.39 '

#### Porqueeee???

### Step 8. How many times was a Veggie Salad Bowl ordered?

In [7]:
df_chipo.value_counts()

order_id  quantity  item_name           choice_description                                                                                                                                                   item_price
926       1         Chicken Bowl        [Roasted Chili Corn Salsa, [Fajita Vegetables, Rice, Cheese, Sour Cream, Lettuce]]                                                                                   $8.75         3
724       1         Canned Soft Drink   [Coke]                                                                                                                                                               $1.25         3
1011      1         Canned Soft Drink   [Coke]                                                                                                                                                               $1.25         3
838       1         Chicken Soft Tacos  [Fresh Tomato Salsa, [Cheese, Sour Cream, Lettuce]]                              

In [8]:
df_chipo["Veggie Salad Bowl"].value_counts()

KeyError: 'Veggie Salad Bowl'

In [9]:
print(df_chipo.count("Veggie Salad Bowl"))

ValueError: No axis named Veggie Salad Bowl for object type DataFrame

In [10]:
df_chipo.set_index("item_name")

order_id  quantity  \
item_name                                                   
Chips and Fresh Tomato Salsa                  1         1   
Izze                                          1         1   
Nantucket Nectar                              1         1   
Chips and Tomatillo-Green Chili Salsa         1         1   
Chicken Bowl                                  2         2   
...                                         ...       ...   
Steak Burrito                              1833         1   
Steak Burrito                              1833         1   
Chicken Salad Bowl                         1834         1   
Chicken Salad Bowl                         1834         1   
Chicken Salad Bowl                         1834         1   

                                                                      choice_description  \
item_name                                                                                  
Chips and Fresh Tomato Salsa                                                         NaN   
Izze                                                                        [Clementine]   
Nantucket Nectar                                                                 [Apple]   
Chips and Tomatillo-Green Chili Salsa                                                NaN   
Chicken Bowl                           [Tomatillo-Red Chili Salsa (Hot), [Black Beans...   
...                                                                                  ...   
Steak Burrito                          [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...   
Steak Burrito                          [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...   
Chicken Salad Bowl                     [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...   
Chicken Salad Bowl                     [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...   
Chicken Salad Bowl                     [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...   

                                      item_price  
item_name                                         
Chips and Fresh Tomato Salsa              $2.39   
Izze                                      $3.39   
Nantucket Nectar                          $3.39   
Chips and Tomatillo-Green Chili Salsa     $2.39   
Chicken Bowl                             $16.98   
...                                          ...  
Steak Burrito                            $11.75   
Steak Burrito                            $11.75   
Chicken Salad Bowl                       $11.25   
Chicken Salad Bowl                        $8.75   
Chicken Salad Bowl                        $8.75   

[4622 rows x 4 columns]

In [11]:
print(df_chipo.count("Veggie Salad Bowl"))

ValueError: No axis named Veggie Salad Bowl for object type DataFrame

In [17]:
df_element= df_chipo[["item_name", "quantity"]]

In [18]:
df_element

item_name  quantity
0              Chips and Fresh Tomato Salsa         1
1                                      Izze         1
2                          Nantucket Nectar         1
3     Chips and Tomatillo-Green Chili Salsa         1
4                              Chicken Bowl         2
...                                     ...       ...
4617                          Steak Burrito         1
4618                          Steak Burrito         1
4619                     Chicken Salad Bowl         1
4620                     Chicken Salad Bowl         1
4621                     Chicken Salad Bowl         1

[4622 rows x 2 columns]

In [24]:
print(df_element.count( "Veggie Salad Bowl"))

ValueError: No axis named Veggie Salad Bowl for object type DataFrame

In [ ]:
df

### Step 9. How many times did someone order more than one Canned Soda?